src/expr/src/scalar/func.rs

```diff
@@ -1244,6 +1244,74 @@ fn cot<'a>(a: Datum<'a>) -> Result<Datum, EvalError> {
     Ok(Datum::from(1.0 / f.tan()))
 }
 
+fn log_guard(val: f64, function_name: &str) -> Result<f64, EvalError> {
+    if val.is_sign_negative() {
+        return Err(EvalError::NegativeOutOfDomain(function_name.to_owned()));
+    }
+    if val == 0.0 {
+        return Err(EvalError::ZeroOutOfDomain(function_name.to_owned()));
+    }
+    Ok(val)
+}
+
+fn log_base<'a>(a: Datum<'a>, b: Datum<'a>, scale: u8) -> Result<Datum<'a>, EvalError> {
+    let b = cast_significand_to_float64(b).unwrap_float64() / 10_f64.powi(i32::from(scale));
```

Comment:
tiny, ignorable nit: the `i32::from(scale)` bit is separated out in the other functions you've added, if you'd want to make it a separate line here as well

Reply:
True! I had considered that but would have to re-convert to a `u8` again, or retain the original; I thought this one read better (: